### PR TITLE
[APID-281] Validate Sidekiq Tasks Timeout Settings

### PR DESCRIPTION
**Pull request by ttr3dp**
**JIRA:** https://digital-science.atlassian.net/browse/APID-281

### Why? ###

> When a job has to be terminated, Nomad first send a `SIGTERM` to the task, waits a number of seconds (5 by default) and then sends a `SIGKILL` if the task has not exited (https://developer.hashicorp.com/nomad/docs/job-specification/task#kill_timeout)
When sidekiq receives the `SIGTERM`, it waits by default 25 seconds. If after 25 seconds there is any job running, sidekiq pushes the jobs back to Redis so they can be rerun later (https://github.com/mperham/sidekiq/wiki/Deployment#overview).
But given that by default the timeout in a Nomad task is smaller than the timeout in sidekiq, jobs that take more than 5s to run can be lost on deploys.

Long story short, if Nomad task is supposed to execute `sidekiq` process, we must pass `-t` argument which is `5` seconds less than `kill_timeout` parameter in the task stanza.

### What's changed? ###

- Add https://github.com/mfcastellani/hcl-checker gem for parsing job into hash
- Add custom `SidekiqTaskValidator` validator - contains logic for checking if the timeout(s) are properly set in Sidekiq tasks **only**
- Add validator tests

**OTHER:**

- Generated `bundler` binstub, so `./bin/bundle` is now available
- Update `nomad` version to `1.4.2`, so it matches our environments

### How it looks like? ###

Screenshot 2023-01-26 at 10 14 32

### CODE REVIEW

File: app/validators/sidekiq_task_validator.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    groups.each do |group|
+      group.each do |group_name, group_spec|
+        tasks = group_spec['task']
+
+        next if tasks.blank?
+
+        # Same as with groups above, we are making sure to always deal with an array.
+        tasks = [tasks] if tasks.is_a?(Hash)
+
+        tasks.each do |task|
```
Comment (PatoSoft) [inline app/validators/sidekiq_task_validator.rb:30]: You can also do `task.to_a.each` directly

**Review by PatoSoft — APPROVED** [review]
Great job 👏 👏 👏 @ttr3dp

File: app/validators/sidekiq_task_validator.rb (hunk 1 above)
Comment (ttr3dp) [reply app/validators/sidekiq_task_validator.rb:30]: I tried that and also `Array(hash)`, but it doesn't work. It does not wrap the hash into array, but makes the key and value elements of it. Check it out:

```rb
[1] pry(main)> hash = { 'task' => { 'config' => 'blah' } }
=> {"task"=>{"config"=>"blah"}}
[2] pry(main)> hash.to_a
=> [["task", {"config"=>"blah"}]]
[3] pry(main)> Array(hash)
=> [["task", {"config"=>"blah"}]]
[4] pry(main)> [hash]
=> [{"task"=>{"config"=>"blah"}}]
```

File: app/validators/sidekiq_task_validator.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    full_command = task['config'].values_at('command', 'args').flatten.join(' ')
+
+    return true unless full_command.include?('sidekiq')
+
+    sidekiq_timeout = full_command[/^* -t (.*) */, 1]
+    kill_timeout = task["kill_timeout"]&.sub('s', '')
+
+    return false if sidekiq_timeout.blank? || kill_timeout.blank?
+
+    (kill_timeout.to_i - sidekiq_timeout.to_i) == 5
```
[the remark this answers is not in the document]
Comment (ttr3dp) [reply app/validators/sidekiq_task_validator.rb:56]: Yep...we can do that... 👍

[the remark this answers is not in the document]
Comment (ttr3dp) [reply app/validators/sidekiq_task_validator.rb]: Done https://github.com/PowerRhino/rocksteady/pull/110/commits/0d1307cc9788184e61b07788156f879243551557
